### ISSUE TITLE
Router: should be able to set the browser history title

### ISSUE THREAD

**Issue by wardbell** [angular/angular#7630]
**I'm submitting a ...**
[ ] bug report
[x] feature request
[ ] support request.

No obvious way to influence the history title (see [this background](https://css-tricks.com/using-the-html5-history-api/)) that appears when you press-and-hold a browser's back/forward buttons.

We either need this as a router feature or have some reasonable technique available for devs to implement themselves.

This title should be dynamic as the displayed value is often constructed on the fly by the component (e.g, "Hero Detail: Magneta"). I'd like to see this option be part of the link parameters array and somehow involved in the `navigateTo` methods.

The A1 routers could do it. So can routers for other frameworks. We need it.

**Comment by mgol** [comment]
This is just the regular `document.title` from before the `pushState`, no?

**Comment by choeller** [comment]
@wardbell I first thought this would be cool, but now  I'm kind of wondering if the link-Syntax / navigateTo-Method really is the right place to decide about the brower title of the target-component. 

Considering having multiple links to the same location you would need to redefine the logic for constructing the title in each link. Also thinking in "encapsulated components" the knowledge of how to construct the title should maybe belong to the component itself. Can you explain why you would like to see it in the link-Syntax? 

Also - did you know about the `Title`-Service? (https://angular.io/docs/ts/latest/api/platform/browser/Title-class.html) Thanks to @ericmartinezr I just found out about it, and in combination with `RouteData` and `router.subscribe` I was able to satisfy all my usecases.

**Comment by choeller** [comment]
If anyone is interested this is how I now solved it in my application. In my main component I have the following:
```ts
  constructor(private router:Router,private titleService: Title) {  }

  ngOnInit() {
    let defaultTitle = this.titleService.getTitle();
    let current = "";
    this.router.subscribe((route) => {
      if (current === route) { return; } //needed to be able to overwrite the title from component
      current = route;
      //following 2 lines are a little hacky, as we need to distinguish beetween top-level and child-routes
      let component:any = this.router.currentInstruction.child || this.router.currentInstruction.component;
      component = component.component || component;
      this.titleService.setTitle(component.routeData.data['title'] || defaultTitle);
    });
  }
```
It would be cool, to have a cleaner way to access the currently active component, but by the lines above I'm now able to set my page title from '@RouteConfig' like this:
```ts
  {path: '/newtask', component: EditTask, name: 'NewTask',  data: {title:"Create new Task"}},
```
or by directly writing to the title-service from my component...

**Comment by wardbell** [comment]
I was pretty sure that the `TitleService` governs the `document.title` that shows in the browser tab label which I think is **not the push state title** that appears in the pop-up when we press-and-hold the browser's forward/back buttons.

Am I wrong about that?

**Comment by zoechi** [comment]
There is a title argument (`history.pushState([data], [title], [url])`) but it seems to be univerally ignored https://developer.mozilla.org/en-US/docs/Web/API/History/pushState, https://css-tricks.com/using-the-html5-history-api/ (1 year old). If an empty string is passed `document.string` seems to be used.

**Comment by wardbell** [comment]
Oh crap!

>the title is the second parameter which can be a string, but at the time of writing, every browser simply ignores it.

But do we agree that the router should have a facility to set the title as part of the navigation? I'm not fond of the trouble @choeller  went through to set the `document.title`. We should make it easy to do in the course of the navigation process, don't you think?

**Comment by mgol** [comment]
> If an empty string is passed document.string seems to be used.

Don't you mean `document.title`? Also, I think the `title` parameter is completely ignored which means `document.title` should always be used?

**Comment by zoechi** [comment]
I'd expect support for a static string value on the route config and alternatively a lifecycle callback that allows to dynamically generate a string just before `pushState()` is executed.

**Comment by choeller** [comment]
@wardbell I totally agree, that we should have an easier option to set the browser title. My point is, that IMO this should _not_ belong to the navigation part (ngRouterLink or navigateTo) but rather to the config of the route. Putting it to the navigation part IMO somehow implies wrong responsibilities, as the component calling the target would decided about the title of the target. I'm not sure if I get your point about the `navigateTo`. When saying something like 
```
class SomeOtherClass {
...
  save() {
  ...
  router.navigateTo(['Overview'], title: 'The Overview');
  }
}
```
you would still let SomeOtherComponent decide for the title of the Overview-Component, or am I not getting the idea?

**Comment by zoechi** [comment]
Another question just occuring: Should titles from parent and child routes be concatenated or should only one component be able to set the title?

**Comment by choeller** [comment]
About the Title-Service: For me it perfectly sets the history push-state. I checked in Chrome and Firefox and both contain the correct title I set via Title-Service. So here is the proof: (Excuse the german in it :))
![bildschirmfoto vom 2016-04-04 11 42 32](https://cloud.githubusercontent.com/assets/423277/14244731/edb98904-fa5c-11e5-8ef2-517d772b634d.png)

**Comment by choeller** [comment]
@zoechi IMO some magic, like concatenating the titles from parent and child lead to more confusion then you benefit from it. So someone wants to have a Seperator in it like `Tasks -> Overview` and someone else just wants to say `Task Overview` in the title. My personal opinion is to introduce as less magic as possible...

**Comment by wardbell** [comment]
@choeller That's what we're looking for. Thanks for that confirmation! What remains to be done is make it easy to set that title during navigation. We should leave this issue open until it's done ... although, given that you have a viable workaround, perhaps we should de-prioritize it relative to the things we must do and can't do presently.

@zoechi  Your question has no answer ... other than "I don't think we should automate this"  I believe it is the developer's job to determine the title for each navigation. And I see that @choeller  just said that a bit more elegantly.

**Comment by choeller** [comment]
@wardbell yeah, so my use-cases are now working as I need them, but I totally agree that having to subscribe to the router and doing some quite wordy calculations is still kind of a PITA. Nevertheless I also see that other topics, especially the DI / LifeCycle Stuff are way more essential for the RC...

**Comment by dheeran1** [comment]
+!

**Comment by atscott** [comment]
Copying from #41626

# 🚀 feature request

### Relevant Package
This feature request is for @angular/packages/router and @angular/angular.


### Description
In order to increase accessibility and build in best practices, allow an option to provide a unique page title to each path in the Angular router. 


#### TitleService vs Proposed Change
Currently the TitleService allows us to manage page title, which allows developers to add unique page titles. This requires additional work for developers and the contextual knowledge to understand that this not only a best practice, but that page title _needs_ to be managed in order to make your application more accessible. 

Eventually I'd like to see this option enforced as mandatory or throw a warning if not defined, to ensure Angular apps are more accessible by default. This known property could also be more easily caught with a lint rule.

#### Context
W3 defines that [page title must be unique](https://www.w3.org/WAI/WCAG21/Understanding/page-titled). Angular is a single-page app and, as a result, a majority of transitions, such as moving to a new page, do not involve a page reload. This means that out of the box, each page is read with an identical page title, and provides no value for understanding the page's content or purpose for users using a11y services.


### Proposed feature
Define a know `title` property in the router's route definitions, which is then subscribed to automatically and uses the TitleService under the hood to manage page title on routing automatically.

```ts
const routes: Routes = [
  { path: 'chicken', component: ChickenComponent, title: 'Chicken - Farm Demo App' },
  { path: 'goat', component: GoatComponent, title: 'Goat - Farm Demo App' },
  { path: 'frog', component: FrogComponent, title: 'Frog - Farm Demo App' },
  { path: '',   redirectTo: '/goat', pathMatch: 'full' }, // redirects ok without a title
  { path: '**', component: HomeComponent }, // throws warning with no title prop defined
];
```

### Describe alternatives you've considered
TitleService works, but again, it puts the burden on the developer to know about this single page app accessibility pitfall and how to implement unique page titles (and manage subscribing to the router ActivatedRoute or manage title in each route's component).

**Comment by IgorMinar** [comment]
Just a few notes from our recent discussion about this issue with the fw subteam and @twerske:

- there is a general support for this feature, it is going to be a great addition
- there are concerns around:
  - i18n
  - dynamic / parametarized titles, and
  - behavior when the routes are nested

These concerns should be resolved as part of the design before we commit to any particular API.

@petebacondarwin mentioned that the i18n concern could be resolved by using `$localize` service:

```
const routes: Routes = [
  { path: 'chicken', component: ChickenComponent, title: $localize`Chicken - Farm Demo App`},
  ...
]
```

For parametrized routes, which is very common  we might need to go with an API similar to that of Resolvers and Guards. Here is a rough sketch:

```
const routes: Routes = [
  { path: 'chickens/:chickenid', component: ChickenComponent, title: ChickenTitle},
  ...
]
```

And for the third concern around nested routes, we need to do more research about what the correct behavior should be.  Concat of title coming from various segments vs the deepest segment's title takes over are just two of the possible solutions.
